src/Evolution/Initialization/FunctionOfTimeFromFile.hpp

```diff
@@ -0,0 +1,164 @@
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "Domain/FunctionsOfTime/FunctionOfTime.hpp"
+#include "Domain/FunctionsOfTime/PiecewisePolynomial.hpp"
+#include "Domain/FunctionsOfTime/Tags.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Evolution/Initialization/Tags.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/Dat.hpp"
+#include "IO/H5/File.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Initialization {
+namespace Actions {
+/// \ingroup InitializationGroup
+/// \brief Initialize a `FunctionsOfTime` from dat files in an
+/// H5 file.
+///
+/// Uses:
+///  - DataBox:
+///    - `Initialization::Tags::FuncOfTimeFile`
+///    - `Initialization::Tags::FuncOfTimeSetNames`
+///
+/// DataBox changes:
+/// - Adds:
+///   - `::domain::Tags::FunctionsOfTime`. This tag contains a `FunctionOfTime`
+///   for each dat file read from the H5 file
+///
+/// - Removes: nothing
+/// - Modifies: nothing
+///
+///
+/// Columns in the file to be read must have the following form:
+///   - 0 = time
+///   - 1 = time of last update
+///   - 2 = number of components
+///   - 3 = maximum derivative order
+///   - 4 = version
+///   - 5 = function
+///   - 6 = d/dt (function)
+///   - 7 = d^2/dt^2 (function)
+///   - 8 = d^3/dt^3 (function)
+///
+/// If the function has more than one component, columns 5-8 give
+/// the first component and its derivatives, columns 9-12 give the second
+/// component and its derivatives, etc.
+///
+struct FunctionOfTimeFromFile {
+  template <
+      typename DbTagsList, typename... InboxTags, typename Metavariables,
+      typename ArrayIndex, typename ActionList, typename ParallelComponent,
+      Requires<
+          db::tag_is_retrievable_v<Initialization::Tags::FuncOfTimeFile,
+                                   db::DataBox<DbTagsList>> and
+          db::tag_is_retrievable_v<Initialization::Tags::FuncOfTimeSetNames,
+                                   db::DataBox<DbTagsList>>> = nullptr>
+  static auto apply(db::DataBox<DbTagsList>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::ConstGlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& file_name{
+        db::get<Initialization::Tags::FuncOfTimeFile>(box)};
+
+    // Each element of dataset_names has two elements:
+    // element 0 is the dataset name in the h5 file,
+    // and element 1 is the name used in the std::unordered_map that
+    // will store the corresponding read-in function of time
+    const std::vector<std::array<std::string, 2>>& dataset_names{
```

Comment:
Because spec might use very un-spectre-like naming conventions. Also, re: your question to @markscheel, I should amend the comments. I now realize that the derive order is not always set to 3 in spec, but sometimes is set to 2 instead. For now, though, I only want to mimic a simple spec control system using max_deriv=3. If we end up wanting to mimic a more complex control system later, then this can be generalized to take max_deriv as an input option, but I wanted to keep the first version of this simple for now

Reply:
make this a std::map<std::string, std::string>